src/util/tmpl/fd_slist.c

```diff
@@ -0,0 +1,476 @@
+   2 - library implementation */
+
+#ifndef SLIST_IMPL_STYLE
+#define SLIST_IMPL_STYLE 0
+#endif
+
+/* Implementation *****************************************************/
+
+/* Constructors and verification log details on failure (rest only needs
+   fd_bits.h, consider making logging a compile time option). */
+
+#include "../log/fd_log.h"
+
+#define SLIST_(n) FD_EXPAND_THEN_CONCAT3(SLIST_NAME,_,n)
+
+#if SLIST_IMPL_STYLE==0 || SLIST_IMPL_STYLE==1 /* need structures and inlines */
+
+struct __attribute__((aligned(8))) SLIST_(private) {
+
+  /* join points here */
+
+  SLIST_IDX_T head;
+  SLIST_IDX_T tail;
+};
+
+typedef struct SLIST_(private) SLIST_(private_t);
+
+typedef SLIST_(private_t) SLIST_(t);
+
+FD_PROTOTYPES_BEGIN
+
+/* slist_private returns the location of the slist header for a current
+   local join.  Assumes join is a current local join.
+   slist_private_const is a const correct version. */
+
+FD_FN_CONST static inline SLIST_(private_t) *
+SLIST_(private)( SLIST_(t) * join ) {
+  return (SLIST_(private_t) *)join;
+}
+
+FD_FN_CONST static inline SLIST_(private_t) const *
+SLIST_(private_const)( SLIST_(t) const * join ) {
+  return (SLIST_(private_t) const *)join;
+}
+
+/* list_private_{cidx,idx} compress / decompress 64-bit in-register
+   indices to/from their in-memory representations. */
+
+FD_FN_CONST static inline SLIST_IDX_T SLIST_(private_cidx)( ulong      idx   ) { return (SLIST_IDX_T)idx;  }
+FD_FN_CONST static inline ulong       SLIST_(private_idx) ( SLIST_IDX_T cidx ) { return (ulong)     cidx; }
+
+/* list_private_idx_null returns the element storage index that
+   represents NULL. */
+
+FD_FN_CONST static inline ulong SLIST_(private_idx_null)( void ) { return (ulong)(SLIST_IDX_T)~0UL; }
+
+/* list_private_idx_is_null returns 1 if idx is the NULL list index
+   and 0 otherwise. */
+
+FD_FN_CONST static inline int SLIST_(private_idx_is_null)( ulong idx ) { return idx==(ulong)(SLIST_IDX_T)~0UL; }
+
+FD_FN_CONST static ulong SLIST_(ele_max)( void ) { return (ulong)(SLIST_IDX_T)~0UL; }
+
+FD_FN_PURE static inline int
+SLIST_(is_empty)( SLIST_(t) const * join ) {
+  return SLIST_(private_idx_is_null)( SLIST_(private_idx)( SLIST_(private_const)( join )->head ) );
+}
+
+FD_FN_PURE static inline ulong
+SLIST_(idx_peek_head)( SLIST_(t) const * join ) {
+  return SLIST_(private_idx)( SLIST_(private_const)( join )->head );
+}
+
+FD_FN_PURE static inline ulong
+SLIST_(idx_peek_tail)( SLIST_(t) const * join ) {
+  return SLIST_(private_idx)( SLIST_(private_const)( join )->tail );
+}
+
+static inline SLIST_(t) *
+SLIST_(idx_push_head)( SLIST_(t) *   join,
+                       ulong         ele_idx,
+                       SLIST_ELE_T * pool ) {
+  SLIST_(private_t) * list = SLIST_(private)( join );
+
+  ulong head_idx = SLIST_(private_idx)( list->head );
+
+  pool[ ele_idx ].SLIST_NEXT = SLIST_(private_cidx)( head_idx );
+
+  if( SLIST_(private_idx_is_null)( head_idx ) ) {
+    list->tail = SLIST_(private_cidx)( ele_idx );
+  }
```

Comment:
Could do branchless with a store if.

Reply:
I'm pretty sure this also compiles to a cmov